pennylane/ops/qubit/parametric_ops_multi_qubit.py

```diff
@@ -604,13 +604,126 @@ def _pauli_rot_resources(pauli_word):
+    :math:`k=121_{10}=01111001_2`, which can be (trivially) decomposed into a sum of
+    five powers of two by reading off the bits:
+    :math:`k = 2^6 + 2^5 + 2^4 + 2^3 + 2^0 = 64 + 32 + 16 + 8 + 1`.
+    The decomposition here, however, allows for minus signs and achieves the decomposition
+    :math:`k = 2^7 - 2^3 + 2^0 = 128 - 8 + 1`, which only requires three powers of two.
+    """
+    R = []
+    assert k <= 2 ** (n - 1)
+    s = 0
+    powers = 2 ** np.arange(n)
+    for p in powers:  # p = 2**(n-1-i)
+        if s & p == k & p:
+            # Equal bit, move on
+            factor = 0
+        else:
+            # Differing bit, consider pairs of bits
+            if p >= 2 ** (n - 2):
+                # 2**(n-1-i) >= 2**(n-2) is the same condition as i < 2
+                factor = 1
+            else:
+                # Table entry from docstring
+                in_middle_rows = (s & (p + 2 * p)).bit_count() == 1  # two bits of s are 01 or 10
+                in_last_cols = bool(k & (2 * p))  # latter bit of k is 1
+                if in_middle_rows != in_last_cols:  # xor between in_middle_rows and in_last_cols
+                    factor = -1
+                else:
+                    factor = 1
+
+            s += factor * p
+        R.insert(0, factor)
+
+    return R
+
+
+def _ctrl_phase_shift(phi, wire, subspace, control_data):
+    r"""Implement a ((multi-)controlled) phase shift on the specified subspace of a
+    target qubit/wire.
+
+    Args:
+        phi (float): Phase shift angle
+        wire (~.Wires): Target wire
+        subspace (int): which subspace of the target wire the phase shift is applied to. 0 or 1.
+        control_data (tuple): A three-tuple containing control wires, control values, and work
+            wires of the ((multi-)controlled) phase shift gate.
+
+    Returns:
+        list[~.Operation]: The operations implementing the phase shift.
+
+    The decomposition for subspace=1 always is a simple ``PhaseShift`` gate, or its controlled
+    counterpart.
+    The decomposition of a non-controlled phase shift for subspace=0 can be achieved in two ways:
+    The first is to flip the angle of the phase shift and complementing it with a global phase,
+    so that the (diagonal of the) gate is decomposed as
+
+    .. math::
+
+        (\exp(i\phi), 1) = (1, \exp(-i\phi)) (\exp(i\phi), \exp(i\phi)).
+
+    The second is to conjugate a phase shift by Pauli-X operators on the same qubit, decomposing
+    the gate matrix as
+
+    / e^(i\phi)   0 \ _ / 0   1 \/ 1     0     \/ 0   1 \
+    \     0       1 / ¯ \ 1   0 /\ 0 e^(i\phi) /\ 1   0 /.
+
+    Without controls, the first approach is nicer, because global phases usually are free
+    operations. With controls, however, this approach would lead to a controlled global
+    phase, which is equivalent to a phase shift on top of the controlled phase shift required
+    anyways. For the second approach, we may use the ctrl(compute-uncompute) pattern to avoid
```

Comment:
since this is not super standard could be worth linking the ctrl(compute-uncompute) reference

Reply:
Pending crowd-sourcing a good reference here.